# Section A: 20 Marks

In [536]:
import numpy as np
import pandas as pd

### 1. Write a Program
#### 1. Create a list of numbers from 1 to 10 in python (do not hardcode numbers). (1 Marks)

In [537]:
y=list(np.arange(1,11))
y

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

#### 2. Find the number of elements present in the list. (1 Marks)

In [538]:
len(y)

10

#### 3. Convert the above list into a series. (1 Marks)

In [1]:
yn=pd.Series(y)
yn

NameError: name 'pd' is not defined

#### 4. Display the elements at even position. (1 Marks)

In [540]:
yn.iloc[::2]

0    1
2    3
4    5
6    7
8    9
dtype: int64

#### 5. Add the elements 11,12 and 13 to the existing series. (1 Marks)

In [541]:
yn.append(pd.Series([11,12,13], index=[10,11,12]))

0      1
1      2
2      3
3      4
4      5
5      6
6      7
7      8
8      9
9     10
10    11
11    12
12    13
dtype: int64

### 2.Read the dataset(Euro_2012_stats_TEAM.CSV)

In [542]:
df=pd.read_csv("C:/Users/Anand/Desktop/Yash/Euro_2012_stats_TEAM.csv")
df

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  \
0              51.9%            16.0%                          32   
1              41.9%            12.9%                          39   
2              50.0%            20.0%                          27   
3              50.0%            17.2%                          40   
4              37.9%             6.5%                          65   
5              47.8%            15.6%                          80   
6              30.7%            19.2%                          32   
7              43.0%             7.5%                         110   
8              25.0%             4.1%                          60   
9              39.4%             5.2%                          48   
10             34.3%             9.3%                          82   
11             36.8%             5.2%                          28   
12             22.5%            12.5%                          59   
13             55.9%            16.0%                         100   
14             47.2%            13.8%                          39   
15             21.2%             6.0%                          38   

    Hit Woodwork  Penalty goals  Penalties not scored      ...       \
0              0              0                     0      ...        
1              0              0                     0      ...        
2              1              0                     0      ...        
3              0              0                     0      ...        
4              1              0                     0      ...        
5              2              1                     0      ...        
6              1              1                     1      ...        
7              2              0                     0      ...        
8              2              0                     0      ...        
9              0              0                     0      ...        
10             6              0                     0      ...        
11             0              0                     0      ...        
12             2              0                     0      ...        
13             0              1                     0      ...        
14             3              0                     0      ...        
15             0              0                     0      ...        

    Saves made  Saves-to-shots ratio  Fouls Won Fouls Conceded  Offsides  \
0           13                 81.3%         41             62         2   
1            9                 60.1%         53             73         8   
2           10                 66.7%         25             38         8   
3           22                 88.1%         43             45         6   
4            6                 54.6%         36             51         5   
5      

#### 1. How many teams participated in the Euro2012 dataset. (1 mark)

In [543]:
len(df["Team"].value_counts())

16

In [544]:
pd.unique(df['Team']).size


16

#### 2. Create a data frame with a name &quot;discipline&quot; consisting of (Team, Yellow Cards and Red Cards) (1 mark)

In [545]:
discipline = pd.DataFrame(df, columns=['Team','Yellow Cards','Red Cards'])
print(discipline)

                   Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
6                Greece             9          1
7                 Italy            16          0
8           Netherlands             5          0
9                Poland             7          1
10             Portugal            12          0
11  Republic of Ireland             6          1
12               Russia             6          0
13                Spain            11          0
14               Sweden             7          0
15              Ukraine             5          0


#### 3. Find average Yellow Cards given per Team (1 mark)

In [546]:
df.groupby('Team')[['Yellow Cards']].mean()

Yellow Cards
Team                             
Croatia                         9
Czech Republic                  7
Denmark                         4
England                         5
France                          6
Germany                         4
Greece                          9
Italy                          16
Netherlands                     5
Poland                          7
Portugal                       12
Republic of Ireland             6
Russia                          6
Spain                          11
Sweden                          7
Ukraine                         5

#### 4. Find which teams scored more than 6 goals(1 mark)

In [547]:
df[(df['Goals'] > 6)]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5   Germany     10               32                32             47.8%   
13    Spain     12               42                33             55.9%   

   % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5             15.6%                          80             2              1   
13            16.0%                         100             0              1   

    Penalties not scored      ...       Saves made  Saves-to-shots ratio  \
5                      0      ...               10                 62.6%   
13                     0      ...               15                 93.8%   

    Fouls Won Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  \
5          63             49        12             4          0       15   
13        102             83        19            11          0       17   

    Subs off  Players Used  
5         15            17  
13        17            18  

[2 rows x 35 columns]

#### 5. Display the teams start with &quot; G &quot;(1 mark)

In [548]:
for j in df["Team"]:
   if j.startswith("G"):
       print(j)

Germany
Greece


### 3. Read the dataset(Match.csv) and Answer below questions

In [549]:
df1=pd.read_csv("C:/Users/Anand/Desktop/Yash/Match.csv")
df1.head()

Match_Id Match_Date  Team_Name_Id  Opponent_Team_Id  Season_Id  \
0    335987  18-Apr-08             2                 1          1   
1    335988  19-Apr-08             4                 3          1   
2    335989  19-Apr-08             6                 5          1   
3    335990  20-Apr-08             7                 2          1   
4    335991  20-Apr-08             1                 8          1   

                                   Venue_Name  Toss_Winner_Id Toss_Decision  \
0                       M Chinnaswamy Stadium               2         field   
1  Punjab Cricket Association Stadium, Mohali               3           bat   
2                            Feroz Shah Kotla               5           bat   
3                            Wankhede Stadium               7           bat   
4                                Eden Gardens               8           bat   

   IS_Superover  IS_Result  Is_DuckWorthLewis    Win_Type  Won_By  \
0             0          1                  0     by runs   140.0   
1             0          1                  0     by runs    33.0   
2             0          1                  0  by wickets     9.0   
3             0          1                  0  by wickets     5.0   
4             0          1                  0  by wickets     5.0   

   Match_Winner_Id  Man_Of_The_Match_Id  First_Umpire_Id  Second_Umpire_Id  \
0              1.0                  2.0              470               477   
1              3.0                 19.0              471               487   
2              6.0                 90.0              472               512   
3              2.0                 11.0              473               476   
4              1.0                  4.0              474               486   

    City_Name Host_Country  
0   Bangalore        India  
1  Chandigarh        India  
2       Delhi        India  
3      Mumbai        India  
4     Kolkata        India

#### 1. Display the no. of matches played in each Season?(1 Mark)

In [550]:
df1.groupby('Season_Id')['Match_Id'].count()

Season_Id
1    58
2    57
3    60
4    73
5    74
6    76
7    60
8    59
9    60
Name: Match_Id, dtype: int64

#### 2. Display the percentage of matches won by runs? (1 Mark)

In [551]:
len(df1[df1['Win_Type']=='by runs'])/len(df1['Win_Type'])*100

45.233968804159446

#### 3. What percentages of matches are won by the toss winners who chose to bat first?( 1 Mark)

In [552]:
bat1= df1[df1['Toss_Decision'] == 'bat']
toss_win = df1[df1['Match_Winner_Id']==df1['Toss_Winner_Id']]
toss_win['Toss_Decision'].count()/df1['Toss_Decision'].count()*100 

50.43327556325823

#### 4. How many matches are hosted at South Africa?( 1 Mark)

In [553]:
len(df1[df1['Host_Country']=='South Africa']) 

57

#### 5. Print the Id of the team which won Season 2.( 1 Mark)

In [554]:
df1[df1['Season_Id']==2].groupby('Match_Winner_Id')['IS_Result'].count() 

Match_Winner_Id
1.0     3
2.0     9
3.0     8
4.0     7
5.0     6
6.0    10
7.0     5
8.0     9
Name: IS_Result, dtype: int64

### 4.Write a Program

In [555]:
import random

#### 1. Create a 5x5 array with random values and find the minimum and maximum values(1 Mark)

In [556]:
r = np.random.random((5,5))
print("Array:")
print(r)
rmin, rmax = r.min(), r.max()
print("Minimum value:",r.min())
print("Maximum value:",r.max())



Array:
[[0.4145006  0.4736184  0.17873848 0.36845521 0.56191165]
 [0.41086421 0.62714148 0.82811542 0.56618626 0.8797677 ]
 [0.34285123 0.00283658 0.56858451 0.45298643 0.07495755]
 [0.95217868 0.25256168 0.90404932 0.54281392 0.82715018]
 [0.92882128 0.31946124 0.94854862 0.22553134 0.97448146]]
Minimum value: 0.0028365846177080822
Maximum value: 0.9744814551515963


#### 2. Add padding (filled with 0&#39;s) around a random array of size 5*5 created in Question number &#39;A&#39; (1 Mark)

In [557]:
print("array:")
print(r)
print("New array:")
r = np.pad(r, pad_width=1, mode='constant', constant_values=0)
print(r)

array:
[[0.4145006  0.4736184  0.17873848 0.36845521 0.56191165]
 [0.41086421 0.62714148 0.82811542 0.56618626 0.8797677 ]
 [0.34285123 0.00283658 0.56858451 0.45298643 0.07495755]
 [0.95217868 0.25256168 0.90404932 0.54281392 0.82715018]
 [0.92882128 0.31946124 0.94854862 0.22553134 0.97448146]]
New array:
[[0.         0.         0.         0.         0.         0.
  0.        ]
 [0.         0.4145006  0.4736184  0.17873848 0.36845521 0.56191165
  0.        ]
 [0.         0.41086421 0.62714148 0.82811542 0.56618626 0.8797677
  0.        ]
 [0.         0.34285123 0.00283658 0.56858451 0.45298643 0.07495755
  0.        ]
 [0.         0.95217868 0.25256168 0.90404932 0.54281392 0.82715018
  0.        ]
 [0.         0.92882128 0.31946124 0.94854862 0.22553134 0.97448146
  0.        ]
 [0.         0.         0.         0.         0.         0.
  0.        ]]


#### 3. Create a zero matrix of size 5x5 and add values 1,2,3,4,5 as diagonal values of that matrix(1 Mark)

In [558]:
d = np.diag([1, 2, 3, 4, 5])
print(d)

[[1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]
 [0 0 0 0 5]]


#### 4. Given a 1D array (1-20), negate all elements which are between 11 and 18 (Both included),in place. (2 Marks)

In [559]:
arrayy=np.arange(1,21)
arrayy=arrayy[(arrayy < 11) | (arrayy > 18)]
print('All elements which are between 11 and 18 is',arrayy)

All elements which are between 11 and 18 is [ 1  2  3  4  5  6  7  8  9 10 19 20]


# Section B: 20 Marks

## 5. Read the Dataset (Ramen-ratings.CSV)

In [560]:
df2=pd.read_csv("C:/Users/Anand/Desktop/Yash/ramen-ratings.csv")
df2.head()

Review #           Brand  \
0      2580       New Touch   
1      2579        Just Way   
2      2578          Nissin   
3      2577         Wei Lih   
4      2576  Ching's Secret   

                                             Variety Style Country  Stars  \
0                          T's Restaurant Tantanmen    Cup   Japan   3.75   
1  Noodles Spicy Hot Sesame Spicy Hot Sesame Guan...  Pack  Taiwan   1.00   
2                      Cup Noodles Chicken Vegetable   Cup     USA   2.25   
3                      GGE Ramen Snack Tomato Flavor  Pack  Taiwan   2.75   
4                                    Singapore Curry  Pack   India   3.75   

  Top Ten  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

#### 1. Identify how many brands the data is captured. (2 Marks)

In [561]:
df2['Brand'].nunique()

355

#### 2. Identify for how many rows the data for column Top Ten exists? (2 Marks)


In [562]:
df2['Top Ten'].notnull().sum()

37

#### 3. Display the data of bowl noodles for countries China and Taiwan. (3 Marks)

In [563]:
df2.loc[(df2['Style']=='Bowl')&((df2['Country']=='China') | (df2['Country']=='Taiwan'))]

Review #             Brand  \
33        2547  Sichuan Guangyou   
65        2515     Uni-President   
77        2503  Sichuan Guangyou   
118       2462               TTL   
141       2439     Uni-President   
152       2428               TTL   
164       2416             Xiuhe   
166       2414       Little Cook   
179       2401    Sichuan Baijia   
195       2385     Uni-President   
205       2375     Uni-President   
208       2372          Liu Quan   
218       2362    Chuan Wei Wang   
222       2358    Sichuan Baijia   
283       2297    Chuan Wei Wang   
307       2273     Uni-President   
572       2008    Sichuan Baijia   
682       1898       Master Kong   
707       1873             Vedan   
760       1820            Ten-In   
768       1812       Master Kong   
803       1777               JML   
804       1776               JML   
805       1775               JML   
807       1773               JML   
809       1771               JML   
810       1770               JML   
845       1735       Master Kong   
853       1727           Wei Lih   
890       1690      Wugudaochang   
...        ...               ...   
1822       758               JML   
1834       746               JML   
1841       739               JML   
1874       706              Unif   
1946       634            Baijia   
1950       630           Ve Wong   
1995       585           Ve Wong   
2022       558          Long Kow   
2057       523         Dragonfly   
2060       520         Nan Hsing   
2062       518         Dragonfly   
2067       513         Dragonfly   
2071       509           Ve Wong   
2073       507         Dragonfly   
2081       499         Dragonfly   
2122       458           Ve Wong   
2124       456           Ve Wong   
2132       448           Ve Wong   
2144       436              Unif   
2148       432           Ve Wong   
2165       415           Ve Wong   
2195       385           Ve Wong   
2202       378           Ve Wong   
2258       322              Unif   
2262       318              Unif   
2264       316              Unif   
2265       315              Unif   
2279       301         Dragonfly   
2287       293             Tiger   
2537        43       Kim Ve Wong   

                                                Variety Style Country  Stars  \
33                          Chongqing Spicy Hot Noodles  Bowl   China   4.00   
65      Man Han Feast Spicy Beef Flavor Instant Noodles  Bowl  Taiwan   5.00   
77          Sweet Potato Instant Noodle Sout-Hot Flavor  Bowl   China   0.00   
118            Sesame Oil Chicken Noodle With Rice Wine  Bowl  Taiwan   5.00   
141   Man Han Feast Super Hot Pot Beef Flavor Instan...  Bowl  Taiwan   5.00   
152            Sesame Oil Chicken With Rice Wine Noodle  Bowl  Taiwan   5.00   
164                             Hot & Sour Beef Noodles  Bowl   China   4.25   
166                 Thailand Green Curry Instant Noodle  Bowl  Taiwan   5.00   
179               Chongqing Noodles Burning Dry Noodles  Bowl   China   5.00   
195    Man Han Feast Chilli Beef Flavor Instant Noodles  Bowl  Taiwan   4.75   
205   Man Han Feast Braised Pork Flavor Instant Noodles  Bowl  Taiwan   5.00   
208                           Instant Spicy Rice Noodle  Bowl   China   4.00   
218   Bowl Instant Noodles Artificial Spicy Beef Flavor  Bowl   China   4.00   
222           Pickled Cabbage Flavor Instant Vermicelli  Bowl   China   4.00   
283   Bowl Instant Noodles Artificial Beef With Saue...  Bowl   China   3.00   
307              A Q Bucket Noodle Korean Kimchi Flavor  Bowl  Taiwan   4.00   
572                 Hot spicy Flavor Instant Vermicelli  Bowl   China   3.50   
682                                   Spicy Fish Noodle  Bowl   China   3.00   
707            Wei Wei A Instant Noodles Chicken Flavor  Bowl  Taiwan   0.00   
760             Veggie King Buddha Jump Instant Noodles  Bowl  Taiwan   1.50   
768                             Mushroom Instant Noodle  Bowl   China   4.25   
803   Superem

#### 4. Store the result in a data frame. Using the data frame find the brands that have ranking more than 4. (3 Marks

In [564]:
df22=df2.loc[(df2['Style']=='Bowl')&((df2['Country']=='China') | (df2['Country']=='Taiwan'))]
df22=pd.DataFrame(df22)
df22

Review #             Brand  \
33        2547  Sichuan Guangyou   
65        2515     Uni-President   
77        2503  Sichuan Guangyou   
118       2462               TTL   
141       2439     Uni-President   
152       2428               TTL   
164       2416             Xiuhe   
166       2414       Little Cook   
179       2401    Sichuan Baijia   
195       2385     Uni-President   
205       2375     Uni-President   
208       2372          Liu Quan   
218       2362    Chuan Wei Wang   
222       2358    Sichuan Baijia   
283       2297    Chuan Wei Wang   
307       2273     Uni-President   
572       2008    Sichuan Baijia   
682       1898       Master Kong   
707       1873             Vedan   
760       1820            Ten-In   
768       1812       Master Kong   
803       1777               JML   
804       1776               JML   
805       1775               JML   
807       1773               JML   
809       1771               JML   
810       1770               JML   
845       1735       Master Kong   
853       1727           Wei Lih   
890       1690      Wugudaochang   
...        ...               ...   
1822       758               JML   
1834       746               JML   
1841       739               JML   
1874       706              Unif   
1946       634            Baijia   
1950       630           Ve Wong   
1995       585           Ve Wong   
2022       558          Long Kow   
2057       523         Dragonfly   
2060       520         Nan Hsing   
2062       518         Dragonfly   
2067       513         Dragonfly   
2071       509           Ve Wong   
2073       507         Dragonfly   
2081       499         Dragonfly   
2122       458           Ve Wong   
2124       456           Ve Wong   
2132       448           Ve Wong   
2144       436              Unif   
2148       432           Ve Wong   
2165       415           Ve Wong   
2195       385           Ve Wong   
2202       378           Ve Wong   
2258       322              Unif   
2262       318              Unif   
2264       316              Unif   
2265       315              Unif   
2279       301         Dragonfly   
2287       293             Tiger   
2537        43       Kim Ve Wong   

                                                Variety Style Country  Stars  \
33                          Chongqing Spicy Hot Noodles  Bowl   China   4.00   
65      Man Han Feast Spicy Beef Flavor Instant Noodles  Bowl  Taiwan   5.00   
77          Sweet Potato Instant Noodle Sout-Hot Flavor  Bowl   China   0.00   
118            Sesame Oil Chicken Noodle With Rice Wine  Bowl  Taiwan   5.00   
141   Man Han Feast Super Hot Pot Beef Flavor Instan...  Bowl  Taiwan   5.00   
152            Sesame Oil Chicken With Rice Wine Noodle  Bowl  Taiwan   5.00   
164                             Hot & Sour Beef Noodles  Bowl   China   4.25   
166                 Thailand Green Curry Instant Noodle  Bowl  Taiwan   5.00   
179               Chongqing Noodles Burning Dry Noodles  Bowl   China   5.00   
195    Man Han Feast Chilli Beef Flavor Instant Noodles  Bowl  Taiwan   4.75   
205   Man Han Feast Braised Pork Flavor Instant Noodles  Bowl  Taiwan   5.00   
208                           Instant Spicy Rice Noodle  Bowl   China   4.00   
218   Bowl Instant Noodles Artificial Spicy Beef Flavor  Bowl   China   4.00   
222           Pickled Cabbage Flavor Instant Vermicelli  Bowl   China   4.00   
283   Bowl Instant Noodles Artificial Beef With Saue...  Bowl   China   3.00   
307              A Q Bucket Noodle Korean Kimchi Flavor  Bowl  Taiwan   4.00   
572                 Hot spicy Flavor Instant Vermicelli  Bowl   China   3.50   
682                                   Spicy Fish Noodle  Bowl   China   3.00   
707            Wei Wei A Instant Noodles Chicken Flavor  Bowl  Taiwan   0.00   
760             Veggie King Buddha Jump Instant Noodles  Bowl  Taiwan   1.50   
768                             Mushroom Instant Noodle  Bowl   China   4.25   
803   Superem

In [565]:
df22.loc[(df22["Stars"]>4)]

Review #           Brand  \
65        2515   Uni-President   
118       2462             TTL   
141       2439   Uni-President   
152       2428             TTL   
164       2416           Xiuhe   
166       2414     Little Cook   
179       2401  Sichuan Baijia   
195       2385   Uni-President   
205       2375   Uni-President   
768       1812     Master Kong   
804       1776             JML   
805       1775             JML   
810       1770             JML   
853       1727         Wei Lih   
896       1684    Wugudaochang   
1033      1547           Vedan   
1414      1166            Unif   
1874       706            Unif   
2057       523       Dragonfly   
2062       518       Dragonfly   
2081       499       Dragonfly   
2279       301       Dragonfly   

                                                Variety Style Country  Stars  \
65      Man Han Feast Spicy Beef Flavor Instant Noodles  Bowl  Taiwan   5.00   
118            Sesame Oil Chicken Noodle With Rice Wine  Bowl  Taiwan   5.00   
141   Man Han Feast Super Hot Pot Beef Flavor Instan...  Bowl  Taiwan   5.00   
152            Sesame Oil Chicken With Rice Wine Noodle  Bowl  Taiwan   5.00   
164                             Hot & Sour Beef Noodles  Bowl   China   4.25   
166                 Thailand Green Curry Instant Noodle  Bowl  Taiwan   5.00   
179               Chongqing Noodles Burning Dry Noodles  Bowl   China   5.00   
195    Man Han Feast Chilli Beef Flavor Instant Noodles  Bowl  Taiwan   4.75   
205   Man Han Feast Braised Pork Flavor Instant Noodles  Bowl  Taiwan   5.00   
768                             Mushroom Instant Noodle  Bowl   China   4.25   
804                Spicy King Bowl Noodle Spicy Chicken  Bowl   China   5.00   
805                   Spicy King Bowl Noodle Spicy Beef  Bowl   China   4.25   
810                   Spicy King Bowl Noodle Spicy Pork  Bowl   China   4.50   
853                                         Jah Jan Men  Bowl  Taiwan   5.00   
896              Sichuan Pepper & Chicken Flavor Noodle  Bowl   China   4.50   
1033                          Volcano Spicy Pork Noodle  Bowl  Taiwan   4.25   
1414            Man Han Feast Sichuan Chilli Eel Flavor  Bowl  Taiwan   5.00   
1874                     Man Han Feast Spring Onion Eel  Bowl  Taiwan   4.25   
2057                       Artificial Hot & Sour Shrimp  Bowl   China   4.50   
2062                                 Artificial Chicken  Bowl   China   4.50   
2081                                    Artificial Beef  Bowl   China   4.75   
2279                       Guilin Style Rice Vermicelli  Bowl   China   4.50   

     Top Ten  
65       NaN  
118      NaN  
141      NaN  
152      NaN  
164      NaN  
166      NaN  
179      NaN  
195      NaN  
205      NaN  
768      NaN  
804      NaN  
805      NaN  
810      NaN  
853      NaN  
896      NaN  
1033     NaN  
1414     NaN  
1874     NaN  
2057     NaN  
2062     NaN  
2081     NaN  
2279     NaN

### 6. Continue with the 5 Question and dataset

#### 1. For Every Country and Style of noodle find the average Stars. Order them by stars highest to lowest. (4 Marks)

In [566]:
df2.groupby(['Country','Style'])['Stars'].mean().sort_values(ascending=False)

Country        Style
Malaysia       Box      5.000000
Japan          Box      5.000000
USA            Bar      5.000000
Brazil         Cup      4.500000
Japan          Tray     4.375000
Sarawak        Pack     4.333333
Malaysia       Bowl     4.281250
Brazil         Pack     4.250000
Indonesia      Box      4.250000
Singapore      Pack     4.210145
Cambodia       Pack     4.200000
Malaysia       Pack     4.156452
Japan          Bowl     4.140278
Singapore      Bowl     4.096154
Indonesia      Pack     4.091346
Myanmar        Cup      4.083333
Hong Kong      Cup      4.039474
Malaysia       Cup      4.011905
Japan          Cup      3.940816
Indonesia      Cup      3.940476
Singapore      Cup      3.925926
Myanmar        Pack     3.909091
Fiji           Pack     3.875000
South Korea    Bowl     3.865809
               Pack     3.857459
Japan          Pack     3.801613
Taiwan         Pack     3.761326
Thailand       Tray     3.750000
United States  Pack     3.750000
Vietnam        Tray   

#### 2. Write a function to return the year in numbers or null if there is no value for the Top Ten column. Use this value to create a new column Top_Year. (5 Marks) Add the column to the dataset. (1 Marks)

In [567]:
df2.head()

Review #           Brand  \
0      2580       New Touch   
1      2579        Just Way   
2      2578          Nissin   
3      2577         Wei Lih   
4      2576  Ching's Secret   

                                             Variety Style Country  Stars  \
0                          T's Restaurant Tantanmen    Cup   Japan   3.75   
1  Noodles Spicy Hot Sesame Spicy Hot Sesame Guan...  Pack  Taiwan   1.00   
2                      Cup Noodles Chicken Vegetable   Cup     USA   2.25   
3                      GGE Ramen Snack Tomato Flavor  Pack  Taiwan   2.75   
4                                    Singapore Curry  Pack   India   3.75   

  Top Ten  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

In [568]:
df2[['Top_Year','#']]=df2['Top Ten'].str.split(' ',expand=True)
df2.Top_Year.value_counts()




2012    9
2014    8
2015    7
2013    7
2016    6
Name: Top_Year, dtype: int64

# Section C: 30 Marks

### 7. Read the Dataset Fifa.csv and Description for the Dataset is given below
ID: unique id for every player
Name: name
Age: age
Nationality: nationality
Overall: overall rating
Potential: potential rating
Club: current club
Value: current market value
Wage: current wage
Special: special
Preferred Foot: left/right
International Reputation: rating on a scale of 5
Weak Foot: rating on a scale of 5
Skill Moves: rating on a scale of 5
Work Rate: attack work rate/defense work rate
Body Type: body type of player
Real Face: true or false
Position: position on the pitch
Jersey Number: jersey number
Joined: joined date
Contract Valid Until contract end date
Height: Height of the player
Weight: Weight of the player

In [569]:
df3=pd.read_csv("C:/Users/Anand/Desktop/Yash/Fifa.csv")
df3.head()

ID          Name  Age Nationality  Overall  Potential  \
0   16   Luis Garc¡a   37       Spain       71         71   
1   41       Iniesta   34       Spain       86         86   
2   80  E. Bel”zo?lu   37      Turkey       79         79   
3  164      G. Pinzi   37       Italy       70         70   
4  657    D. Vaughan   35       Wales       66         66   

                    Club   Value  Wage  Special   ...   Skill Moves  \
0              KAS Eupen   ?750K   ?6K     1906   ...           3.0   
1            Vissel Kobe  ?21.5M  ?21K     2058   ...           4.0   
2  Medipol Ba?ak?ehir FK     ?4M  ?23K     2047   ...           4.0   
3                 Padova   ?240K   ?2K     1882   ...           3.0   
4           Notts County   ?150K   ?4K     1781   ...           2.0   

        Work Rate  Body Type  Real Face Position Jersey Number     Joined  \
0  Medium/ Medium       Lean         No      RCM          10.0  19-Jul-14   
1    High/ Medium     Normal        Yes       LF           8.0  16-Jul-18   
2  Medium/ Medium     Stocky         No       CM           5.0   9-Jul-15   
3       Low/ High     Normal         No      LCM          11.0  31-Aug-17   
4    Medium/ High     Stocky         No      CDM           8.0   6-Jul-18   

  Contract Valid Until  Height  Weight  
0                 2019    5'10  143lbs  
1                 2021     5'7  150lbs  
2                 2019     5'7  159lbs  
3                 2019    5'11  168lbs  
4                 2019     5'6  154lbs  

[5 rows x 23 columns]

#### 1. Data Manipulation for Value Column (5 Marks).
Step1: Clean the Value column by removing &#39;?&#39; from each row.
Step2: Convert the column in uniform unit i.e. &#39;K&#39; by using given conversion rate 1 =1000 K and also
remove the letter &#39;K&#39; and &#39;M&#39; from each row.
Step3: Convert the Datatype of the column to &#39;float&#39;
First 5 values of Sample output column should look like this:

0      750.0
1    21500.0
2     4000.0

3      240.0
4      150.0
Name: Value, dtype: float64

In [575]:
df3['Value']=df3['Value'].str.replace('?','')
df3['Value']=df3['Value'].str.replace('K','')
df3['Value']=df3['Value'].str.replace('.','')
df3['Value']=df3['Value'].str.replace('M','000')
df3['Value']=df3['Value'].astype(float)

df3['Value'].head()

0       750.0
1    215000.0
2      4000.0
3       240.0
4       150.0
Name: Value, dtype: float64

#### 2. Data Manipulation for Height Column (5 Marks).
Step1: Convert Height column from Feet’ Inches unit to Inches unit by using a given conversion rate
1 feet = 12 inches 
Step2: Convert the Datatype of the Height column to &#39;int&#39;
First 5 values of Sample Output Height column should look like this:

0  61
1   67
2 67
3 61
4 66
Name: Height, dtype: int32

In [570]:
lis1_height = []
for i in df3.Height:
  i = int(i[0])*12+ int(i[2])
  lis1_height.append(i)
    

df3['height']= lis1_height
df3['height'].head()

0    61
1    67
2    67
3    61
4    66
Name: height, dtype: int64

#### 3. Data Manipulation for Weight Column (5 Marks).
Step1: Clean Weight column by removing ‘lbs’ from each row’
Step2: Convert the Datatype of the Weight column to &#39;int&#39;
First 5 values of Sample Output Weight column should look like this:

0    143
1    150
2    159
3    168
4    154
Name: Weight, dtype: int32

In [571]:
weight = []
for i in df3.Weight.str[0:3]:
  weight.append(float(i))
df3['Weight'] = weight
df3['Weight'].head()

0    143.0
1    150.0
2    159.0
3    168.0
4    154.0
Name: Weight, dtype: float64

### 8. Follow the 7 Question and answer the below questions:

#### 1. Find how many different countries players data is included in the dataset (1 Mark)

In [572]:
df3.groupby('Nationality')['Name'].count()

Nationality
Afghanistan               4
Albania                  36
Algeria                  55
Andorra                   1
Angola                   12
Antigua & Barbuda         4
Argentina               707
Armenia                   9
Australia               200
Austria                 282
Azerbaijan                5
Barbados                  3
Belarus                   4
Belgium                 228
Belize                    1
Benin                    15
Bermuda                   1
Bolivia                   6
Bosnia Herzegovina       57
Botswana                  1
Brazil                  689
Bulgaria                 12
Burkina Faso             14
Burundi                   3
Cameroon                 85
Canada                   51
Cape Verde               17
Central African Rep.      3
Chad                      2
Chile                   287
                       ... 
Senegal                  96
Serbia                  109
Sierra Leone              6
Slovakia                 50
Slovenia

#### 2. Find out highest Valued player from the dataset (1 Mark)

In [573]:
df3[df3['Value'] == df3['Value'].max()]['Name']

195                  Pepe
562           R. Eremenko
834          L. Fabia?ski
1021         J. Rodriguez
1050        J. dos Santos
1066             Gervinho
1223           J. Tomkins
1256           F. Caicedo
1301            I. Piatti
1786           A. Carroll
1838           M. Sissoko
1860            M. Parolo
1881     E. Choupo-Moting
2075              A. Poli
2078            O. Ighalo
2132            M. Fabi n
2139           M. Ritchie
2380             S. Hanni
2529             D. Latza
2582               Guaita
2597          A. Prijovi?
2763             B. Andr‚
2814              G. T”re
2925        A. Bertolacci
2966            J. Aquino
3041       Carlos Eduardo
3110          Jaume Costa
3279              A. Gray
3346             F. Kainz
3371            C. Austin
               ...       
6929            S. Fofana
7003          Carlos Man‚
7082          T. Bongonda
7214           G. Nkoudou
7258              Embarba
7285          H. Hateboer
7315            O. Romero
7504     D. 

#### 3. Find which country&#39;s players data is maximum in the dataset and what’s is the count of it (2 Marks)

In [574]:
df3.groupby('Nationality')['Name'].agg(['max','count'])

max  count
Nationality                                    
Afghanistan                     N. Husin      4
Albania                      Y. Ramadani     36
Algeria                        Z. Ferhat     55
Andorra                       Marc Vales      1
Angola                              Rudy     12
Antigua & Barbuda              R. Browne      4
Argentina                       W. P‚rez    707
Armenia                     Y. Movsisyan      9
Australia                     Y. Mustafa    200
Austria                     Z. Junuzovi?    282
Azerbaijan                   R. Dadashov      5
Barbados                       K. Pearce      3
Belarus                       S. Chernik      4
Belgium                          Z. Gano    228
Belize                        M. Salazar      1
Benin                          S. Mouni‚     15
Bermuda                         R. Lambe      1
Bolivia                  J. Garcia Sossa      6
Bosnia Herzegovina             Z. Kozulj     57
Botswana                     R. Batlokwa      1
Brazil                      verton Luiz    689
Bulgaria                        S. Popov     12
Burkina Faso                   Z. Sanogo     14
Burundi                      S. Berahino      3
Cameroon                      Y. Songo'o     85
Canada                Z. Brault-Guillard     51
Cape Verde                       Z‚ Lu¡s     17
Central African Rep.        G. Kondogbia      3
Chad                    M. Mouandilmadji      2
Chile                            Y. Urra    287
...                                  ...    ...
Senegal                         Z. Tour‚     96
Serbia                       Z. Udovi?i?    109
Sierra Leone                  U. Bangura      6
Slovakia                      T. Hubo?an     50
Slovenia                  Z. Ljubijanki?     48
South Africa                   X. Mlambo     55
South Sudan                     K. Athiu      1
Spain                          Zubiaurre    887
St Kitts Nevis                R. Sawyers      3
Sudan                         S. Mustafa      3
Suriname                        M. Abena      4
Sweden                      Z. Jovanovic    354
Switzerland                     Y. Toure    198
Syria                           Y. Kalfa      9
Tanzania                     S. Chilunda      3
Thailand                       W. Jakkit      5
Togo                           S. Lawson     11
Trinidad & Tobago              T. Warner      3
Tunisia                      Y. Touzghar     29
Turkey                          Z. Yavru    274
Uganda                         M. Opondo      6
Ukraine                     Y. Yakovenko     72
United Arab Emirates      O. Abdulrahman      1
United States                Z. Valentin    299
Uruguay                      Y. Quintana    109
Uzbekistan                   O. Akhmedov      2
Venezuela                     Y. Soteldo     41
Wales                       W. Hennessey     93
Zambia                          S. Sunzu      9
Zimbabwe                     W. Katsande     13

[159 rows x 2 columns]

#### 4. Display top 10 clubs ordered by sum of all player’s current market value of that club (3 Marks)

In [576]:
df3.groupby('Club')[['Value']].sum().reset_index().sort_values('Value', ascending=False).head(10)

Club      Value
471          Real Madrid  4590600.0
213         FC Barcelona  4371800.0
375      Manchester City  3789855.0
436  Paris Saint-Germain  3370325.0
316                Inter  3294785.0
584    Tottenham Hotspur  3134175.0
364            Liverpool  2971155.0
327             Juventus  2928575.0
376    Manchester United  2834375.0
135              Chelsea  2386015.0

#### 5. Count the number of players having International Reputation ratings = 4 from the clubs &#39;Real Madrid’ and ‘FC Barcelona’ (3 Marks)

In [578]:
fb = df3[(df3['Club'] == 'Real Madrid') | (df3['Club'] == 'FC Barcelona')]
fb[fb['International Reputation'] == 4]

ID             Name  Age Nationality  Overall  Potential  \
487   152729            Piqu‚   31       Spain       87         87   
549   155862     Sergio Ramos   32       Spain       91         91   
848   165153       K. Benzema   30      France       85         85   
953   168651       I. Rakiti?   30     Croatia       87         87   
1232  173731          G. Bale   28       Wales       88         88   
1364  177003        L. Modri?   32     Croatia       91         91   
1698  181872         A. Vidal   31       Chile       85         85   
1766  182521         T. Kroos   28     Germany       90         90   
2556  189511  Sergio Busquets   29       Spain       89         89   
2930  192119      T. Courtois   26     Belgium       89         90   

              Club     Value   Wage  Special  ...         Work Rate  \
487   FC Barcelona   34000.0  ?240K     1950  ...      High/ Medium   
549    Real Madrid   51000.0  ?380K     2201  ...      High/ Medium   
848    Real Madrid   37000.0  ?240K     2000  ...       Medium/ Low   
953   FC Barcelona  465000.0  ?260K     2184  ...        High/ High   
1232   Real Madrid   60000.0  ?355K     2279  ...      High/ Medium   
1364   Real Madrid   67000.0  ?420K     2280  ...        High/ High   
1698  FC Barcelona   26000.0  ?205K     2230  ...    Medium/ Medium   
1766   Real Madrid  765000.0  ?355K     2190  ...    Medium/ Medium   
2556  FC Barcelona  515000.0  ?315K     2065  ...    Medium/ Medium   
2930   Real Madrid  535000.0  ?240K     1311  ...    Medium/ Medium   

      Body Type  Real Face  Position Jersey Number     Joined  \
487      Normal        Yes       RCB           3.0   1-Jul-08   
549      Normal        Yes       RCB          15.0   1-Aug-05   
848      Normal        Yes        ST           9.0   9-Jul-09   
953      Normal        Yes       RCM           4.0   1-Jul-14   
1232       Lean        Yes        ST          11.0   2-Sep-13   
1364       Lean        Yes       RCM          10.0   1-Aug-12   
1698     Normal        Yes       CAM           8.0   6-Aug-18   
1766     Normal        Yes       LCM           8.0  17-Jul-14   
2556       Lean        Yes       CDM           5.0   1-Sep-08   
2930   Courtois        Yes        GK           1.0   9-Aug-18   

     Contract Valid Until Height  Weight height  
487                  2022    6'4   187.0     76  
549                  2020    6'0   181.0     72  
848                  2021    6'1   179.0     73  
953                  2021    6'0   172.0     72  
1232                 2022    6'1   181.0     73  
1364                 2020    5'8   146.0     68  
1698                 2021   5'11   165.0     61  
1766                 2022    6'0   168.0     72  
2556                 2023    6'2   168.0     74  
2930                 2024    6'6   212.0     78  

[10 rows x 24 columns]

#### 6. Find out the names of the youngest and oldest players from the dataset (5 Marks)

In [577]:
# youngest player
print("Youngest_player:", df3.sort_values('Age').head(1)['Name'].values[0])

# eldest player
print("Eldest_player:", df3.sort_values('Age', ascending=False).head(1)['Name'].values[0])

Youngest player: T. Gundelund
Eldest player: K. Pilkington
